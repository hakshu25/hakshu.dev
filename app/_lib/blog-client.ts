import { createClient } from 'microcms-js-sdk';

const apiKey = process.env.CMS_API_KEY as string;

interface Image {
  url: string;
  height: number;
  width: number;
}

export interface Post {
  id: string;
  title: string;
  body: string;
  image: Image;
  category: string[];
  publishedAt: string;
  revisedAt: string;
  createdAt: string;
  updatedAt: string;
}

export const blogClient = createClient({
  serviceDomain: 'hakshu-blog',
  apiKey: apiKey,
});
