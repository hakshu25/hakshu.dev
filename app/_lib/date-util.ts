export const formatUtcDateTimeToJstDate = (dateTime: string): string => {
  const date = new Date(dateTime);
  const jstDateTime = date.toLocaleDateString('ja-JP');
  if (jstDateTime === 'Invalid Date') {
    return '';
  }

  return jstDateTime;
};
