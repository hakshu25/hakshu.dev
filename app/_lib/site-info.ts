export const siteTitle = "Hakshu's Blog";
