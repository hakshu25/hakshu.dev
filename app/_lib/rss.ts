import { Feed } from 'feed';
import { Post } from './blog-client';

export function generateRssXml(posts: Post[], siteUrl: string) {
  const feed = new Feed({
    title: 'タイトル',
    description: '説明',
    id: siteUrl,
    link: siteUrl,
    language: 'ja',
    favicon: `${siteUrl}/favicon.ico`,
    copyright: '© 2022 hakshu',
    author: {
      name: 'hakshu',
      link: siteUrl,
    },
  });

  posts.forEach((post) => {
    feed.addItem({
      title: post.title,
      id: `${siteUrl}/${post.id}`,
      link: `${siteUrl}/${post.id}`,
      description: post.title,
      content: post.contents,
      date: new Date(post.publishedAt),
      image: post.image.url,
    });
  });

  return feed.rss2();
}
