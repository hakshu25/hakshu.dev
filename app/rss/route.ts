import { SITE_URL } from '../_config';
import { Contents, blogClient } from '../_lib/blog-client';
import { generateRssXml } from '../_lib/rss';

export async function GET() {
  const { contents } = await blogClient.get<Contents>({ endpoint: 'posts' });
  const xml = generateRssXml(contents, SITE_URL);
  return new Response(xml, {
    status: 200,
    headers: {
      'Cache-Control': 's-maxage=86400, stale-while-revalidate', // 24 時間キャッシュ
      'Content-Type': 'text/xml',
    },
  });
}
