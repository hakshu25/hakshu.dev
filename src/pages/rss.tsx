import { GetServerSidePropsContext } from 'next';
import { blogClient } from '../lib/blog-client';
import { generateRssXml } from '../lib/rss';
import { siteUrl } from '../lib/site-info';

export const getServerSideProps = async ({
  res,
}: GetServerSidePropsContext) => {
  const { contents } = await blogClient.get({ endpoint: 'posts' });
  const xml = generateRssXml(contents, siteUrl);

  res.statusCode = 200;
  res.setHeader('Cache-Control', 's-maxage=86400, stale-while-revalidate'); // 24 時間キャッシュ
  res.setHeader('Content-Type', 'text/xml');
  res.end(xml);

  return {
    props: {},
  };
};

const Page = () => null;
export default Page;
