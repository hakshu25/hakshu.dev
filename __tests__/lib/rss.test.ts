import { Post } from '../../app/_lib/blog-client';
import { generateRssXml } from '../../app/_lib/rss';

describe('Rss', () => {
  beforeAll(() => {
    jest
      .useFakeTimers('modern')
      .setSystemTime(new Date('2022-09-25T08:00:00.000Z'));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  describe('generateRssXml()', () => {
    it('フィードを生成する', () => {
      const siteUrl = 'http://localhost:3000/test';
      const posts: Post[] = [
        {
          id: '2sef2d',
          title: 'ブログ1',
          body: 'ブログ1の中身',
          image: {
            url: 'http://localhost:3000/image1',
            height: 300,
            width: 200,
          },
          category: ['Tech'],
          publishedAt: '2022-06-25T08:13:48.183Z',
          revisedAt: '2022-06-28T08:13:48.184Z',
          createdAt: '2022-06-24T08:13:48.183Z',
          updatedAt: '2022-06-28T08:13:48.183Z',
        },
        {
          id: '3ldjsf',
          title: 'ブログ2',
          body: 'ブログ2の中身',
          image: {
            url: 'http://localhost:3000/image2',
            height: 500,
            width: 600,
          },
          category: ['Tips'],
          publishedAt: '2022-08-25T08:13:48.183Z',
          revisedAt: '2022-08-28T08:13:48.184Z',
          createdAt: '2022-08-24T08:13:48.183Z',
          updatedAt: '2022-08-28T08:13:48.183Z',
        },
      ];

      const xml = generateRssXml(posts, siteUrl);

      expect(xml).toMatchSnapshot();
    });
  });
});
