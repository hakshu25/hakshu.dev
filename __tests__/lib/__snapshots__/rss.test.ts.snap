// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rss > generateRssXml() > フィードを生成する 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/">
    <channel>
        <title>タイトル</title>
        <link>http://localhost:3000/test</link>
        <description>説明</description>
        <lastBuildDate>Sun, 25 Sep 2022 08:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
        <language>ja</language>
        <copyright>© 2022 hakshu</copyright>
        <item>
            <title><![CDATA[ブログ1]]></title>
            <link>http://localhost:3000/test/2sef2d</link>
            <guid isPermaLink="false">http://localhost:3000/test/2sef2d</guid>
            <pubDate>Sat, 25 Jun 2022 08:13:48 GMT</pubDate>
            <description><![CDATA[ブログ1]]></description>
            <content:encoded><![CDATA[ブログ1の中身]]></content:encoded>
            <enclosure url="http://localhost:3000/image1" length="0" type="image//image1"/>
        </item>
        <item>
            <title><![CDATA[ブログ2]]></title>
            <link>http://localhost:3000/test/3ldjsf</link>
            <guid isPermaLink="false">http://localhost:3000/test/3ldjsf</guid>
            <pubDate>Thu, 25 Aug 2022 08:13:48 GMT</pubDate>
            <description><![CDATA[ブログ2]]></description>
            <content:encoded><![CDATA[ブログ2の中身]]></content:encoded>
            <enclosure url="http://localhost:3000/image2" length="0" type="image//image2"/>
        </item>
    </channel>
</rss>"
`;
